R/git.R

```diff
@@ -0,0 +1,84 @@
+#' Check if git repo is clean
+#'
+#' Check that git repo is up to date with remote by checking
+#' \code{git status --porcelian}. If import_dir is not a git root directory
+#' this will search parents of the import_dir for the git root
+#'
+#' @param import_dir The directory of the dettl import to check.
+#' @return True if repo is up to date. False otherwise.
+#'
+#' @keywords internal
+git_is_clean <- function(import_dir) {
+  root_dir <- git_root_directory(import_dir)
+  git_status(root_dir)$clean
+}
+
+#' Check git status
+#'
+#' Check that git status using \code{git status --porcelian}. Optionally
+#' ignoring untracked files.
+#'
+#' @param root The git project root directory.
+#' @param ignore_untracked Whether untracked files should be ignored or not.
+#' Passed as \code{--untracked-files} arg to \code{git status}
+#' @return Result from git including output and indication of success.
+#'
+#' @keywords internal
+git_status <- function(root = NULL, ignore_untracked = FALSE) {
+  args <- c("status", "--porcelain",
+            if (ignore_untracked) "--untracked-files=no")
+  res <- git_run(args, root = root, check = TRUE)
+  res$clean <- length(res$output) == 0L
+  res
+}
+
+#' Run a git system command
+#'
+#' Check that git status using \code{git status --porcelian}. Optionally
+#' ignoring untracked files.
+#'
+#' @param args The args to be passed to git
+#' @param root The git project root directory.
+#' @param check Whether exit code should be checked. If exist code not
+#' successful then throw an error.
+#' @return Result from git including output and indication of success.
+#'
+#' @keywords internal
+#'
+#' @examples
+#' dettl:::git_run(c("status", "--porcelain"))
+git_run <- function(args, root = NULL, check = FALSE) {
+  git <- sys_which("git")
+  if (!is.null(root)) {
+    args <- c("-C", root, args)
+  }
+  res <- system3(git, args)
+  if (check && !res$success) {
+    stop(sprintf("Error code %d running command:\n%s",
+                 res$code, paste0("  > ", res$output, collapse = "\n")))
+  }
+  res
+}
+
+#' Locate the git project root directory
+#'
+#' From a subdirectory within a git project locate the root directory.
+#' This is the directory contianing the .git files. Located using
+#' \code{git rev-parse --git-dir}.
+#'
+#' @param dir The directory to locate the git project root from.
+#' @return The git project root normalized, or throws an error if this can't
+#' be located.
+#'
+#' @keywords internal
+git_root_directory <- function(dir) {
```

Comment:
The other way of doing this without resorting to shelling out is to do:

```r
git_root_directory <- function(dir) {
  find_file_descend(".git", dir)
}
```

which should work reasonably well

Reply:
I did think about this but thought that it would be more robust locating the directory using gits tooling. Though taking a look that is a much slower approach (however I doubt those ms are that important here). I guess shelling out is expensive!

Reply:
I'm going to update that. Just did a quick benchmark and using shell approach is 184 times slower on average.